core/drivers/scmi-msg/clock_generic.c

```diff
@@ -0,0 +1,225 @@
+				    size_t *nb_elts)
+{
+	struct scmi_clk *clk = NULL;
+
+	clk = clk_scmi_get_by_id(channel_id, scmi_id);
+	if (!clk)
+		return SCMI_DENIED;
+
+	if (clk_get_rates_array(clk->clk, start_index, rates, nb_elts))
+		return SCMI_GENERIC_ERROR;
+
+	return SCMI_SUCCESS;
+}
+
+unsigned long plat_scmi_clock_get_rate(unsigned int channel_id,
+				       unsigned int scmi_id)
+{
+	struct scmi_clk *clk = NULL;
+
+	clk = clk_scmi_get_by_id(channel_id, scmi_id);
+	if (!clk)
+		return 0;
+
+	return clk_get_rate(clk->clk);
+}
+
+int32_t plat_scmi_clock_set_rate(unsigned int channel_id,
+				 unsigned int scmi_id,
+				 unsigned long rate)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct scmi_clk *clk = NULL;
+
+	clk = clk_scmi_get_by_id(channel_id, scmi_id);
+	if (!clk)
+		return SCMI_DENIED;
+
+	res = clk_set_rate(clk->clk, rate);
+	if (res)
+		return SCMI_GENERIC_ERROR;
+
+	return SCMI_SUCCESS;
+}
+
+int32_t plat_scmi_clock_get_state(unsigned int channel_id,
+				  unsigned int scmi_id)
+{
+	struct scmi_clk *clk = NULL;
+
+	clk = clk_scmi_get_by_id(channel_id, scmi_id);
+	if (!clk)
+		return false;
+
+	return clk->enabled;
+}
+
+int32_t plat_scmi_clock_set_state(unsigned int channel_id,
+				  unsigned int scmi_id,
+				  bool enable_not_disable)
+{
+	struct scmi_clk *clk = NULL;
+
+	clk = clk_scmi_get_by_id(channel_id, scmi_id);
+	if (!clk) {
+		if (enable_not_disable)
+			return SCMI_DENIED;
+		else
+			return SCMI_SUCCESS;
+	}
+
+	if (enable_not_disable) {
+		if (!clk->enabled) {
+			if (clk_enable(clk->clk))
+				return SCMI_GENERIC_ERROR;
+			clk->enabled = true;
+		}
+	} else {
+		if (clk->enabled) {
+			clk_disable(clk->clk);
+			clk->enabled = false;
+		}
+	}
+
+	return SCMI_SUCCESS;
+}
+
+static TEE_Result clk_check_scmi_id(struct clk *new_clk,
+				    unsigned int channel_id,
+				    unsigned int scmi_id)
+{
+	struct scmi_clk *clk = NULL;
+
+	SLIST_FOREACH(clk, &scmi_clk_list, link) {
+		if (clk->channel_id == channel_id && clk->scmi_id == scmi_id) {
+			EMSG("SCMI channel %u, clock %u already registered",
+			     channel_id, scmi_id);
+			return TEE_ERROR_BAD_PARAMETERS;
+		}
+	}
+
+	if (strlen(clk_get_name(new_clk)) >= SCMI_MAX_CLK_NAME_LEN)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result scmi_clk_add(struct clk *clk, unsigned int channel_id,
+			unsigned int scmi_id)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct scmi_clk *scmi_clk = NULL;
+
+	if (scmi_clk_init_done)
+		return TEE_ERROR_BAD_STATE;
+
+	res = clk_check_scmi_id(clk, channel_id, scmi_id);
+	if (res)
+		return res;
+
+	scmi_clk = calloc(1, sizeof(*scmi_clk));
+	if (!scmi_clk)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	scmi_clk->clk = clk;
+	scmi_clk->channel_id = channel_id;
+	scmi_clk->scmi_id = scmi_id;
+	scmi_clk->enabled = 0;
+
+	SLIST_INSERT_HEAD(&scmi_clk_list, scmi_clk, link);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result scmi_clk_init_fini(void)
+{
+	scmi_clk_init_done = true;
+
+	return TEE_SUCCESS;
+}
+
+boot_final(scmi_clk_init_fini);
```

Comment:
I missed that one: prefer `release_init_resource()` as `boot_final()` is not an OP-TEE init final boot stage (when `CFG_NS_VIRTUALIZATION=y`).

Reply:
Acked